hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnectionManagerImpl.java

```diff
@@ -901,32 +904,53 @@ public void run() {
 
     }
 
-    private Collection<Address> getPossibleMemberAddresses() {
-        final List<Address> addresses = new LinkedList<Address>();
+    Collection<Address> getPossibleMemberAddresses() {
+        LinkedHashSet<Address> addresses = new LinkedHashSet<Address>();
 
         Collection<Member> memberList = client.getClientClusterService().getMemberList();
         for (Member member : memberList) {
             addresses.add(member.getAddress());
         }
 
         if (shuffleMemberList) {
-            Collections.shuffle(addresses);
+            shuffle(addresses);
         }
 
-        List<Address> providerAddresses = new ArrayList<Address>();
+        LinkedHashSet<Address> providerAddresses = new LinkedHashSet<Address>();
         for (AddressProvider addressProvider : addressProviders) {
             providerAddresses.addAll(addressProvider.loadAddresses());
         }
 
         if (shuffleMemberList) {
-            Collections.shuffle(providerAddresses);
+            shuffle(providerAddresses);
         }
 
         addresses.addAll(providerAddresses);
 
+        if (previousOwnerConnectionAddress != null) {
+            putPreviousOwnerAddressToLast(addresses);
+        }
         return addresses;
     }
 
+    private static <T> Set<T> shuffle(Set<T> set) {
+        List<T> shuffleMe = new ArrayList<T>(set);
+        Collections.shuffle(shuffleMe);
+        Set<T> shuffledSet = new LinkedHashSet<T>();
+        shuffledSet.addAll(shuffleMe);
+        return shuffledSet;
+    }
+
+    /**
+     * Previous owner address is moved to last item in set so that client will not try to connect to same one immediately.
+     * It could be the case that address is removed because it is healthy(it not responding to heartbeat/pings)
+     * In that case, trying other addresses first to upgrade make more sense.
+     */
+    private void putPreviousOwnerAddressToLast(Set<Address> addresses) {
```

Comment:
changing method signature to accept `LinkedHashSet` instead of `Set` seems more appropriate if order is important. `Set` contract doesn't have an ordering guarantee per se.

Reply:
It was `LinkedHashSet`, but checkstyle complained. :(

Reply:
then maybe something like this: `assert addresses instanceOf LinkedHashSet`

Reply:
It's just a private method, so I think it's okay to enforce a `LinkedHashSet` if it's needed. We should not have this in "public" APIs though.

Maybe in this case a suppression via `@SuppressWarnings("checkstyle:...")` is better, since the `assert` will just be checked at runtime (and will decrease the code coverage, if not tested).

Reply:
I would probably remove the method completely instead. It is just two lines anyway.